bigtable/beam/change-streams/src/main/java/SongRank.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.Top;
+import org.apache.beam.sdk.transforms.windowing.AfterFirst;
+import org.apache.beam.sdk.transforms.windowing.AfterPane;
+import org.apache.beam.sdk.transforms.windowing.AfterProcessingTime;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
+import org.apache.beam.sdk.transforms.windowing.Repeatedly;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+public class SongRank {
+
+  public static void main(String[] args) {
+    BigtableOptions options =
+        PipelineOptionsFactory.fromArgs(args).withValidation().as(
+            BigtableOptions.class);
+    Pipeline p = Pipeline.create(options);
+
+    p.apply(
+            "Stream from Bigtable",
+            BigtableIO.readChangeStream()
+                .withProjectId(options.getBigtableProjectId())
+                .withInstanceId(options.getBigtableInstanceId())
+                .withTableId(options.getBigtableTableId())
+                .withAppProfileId(options.getBigtableAppProfile())
+                .withHeartbeatDuration(Duration.standardSeconds(1))
+        )
+        .apply("Add key", ParDo.of(new ExtractSongName()))
+        .apply(
+            "Collect listens in 5 second windows",
+            Window.<String>into(new GlobalWindows())
+                .triggering(
+                    Repeatedly.forever(
+                        AfterProcessingTime
+                            .pastFirstElementInPane()
+                            .plusDelayOf(Duration.standardSeconds(10))
+                    ))
+                .discardingFiredPanes())
+        .apply(Count.perElement())
+        .apply("Top songs", Top.of(5, new SongComparator()).withoutDefaults())
+        .apply("Print", ParDo.of(new PrintFn()))
+        .apply(
+            "Collect at least 10 elements or 1 minute of elements",
+            Window.<String>into(new GlobalWindows())
+                .triggering(
+                    Repeatedly.forever(
+                        AfterFirst.of(
+                            AfterPane.elementCountAtLeast(10),
+                            AfterProcessingTime
+                                .pastFirstElementInPane()
+                                .plusDelayOf(Duration.standardMinutes(1)
+                                )
+                        )
+                    ))
+                .discardingFiredPanes())
+        .apply(
+            "Output top songs",
+            TextIO.write()
+                .to(options.getOutputLocation() + "song-charts/")
+                .withSuffix(".txt")
+                .withNumShards(1)
+                .withWindowedWrites()
+        );
+
+    p.run();
```

Comment:
```suggestion
    p.run().waitUntilFinish();
```

You'd need `.waitUntilFinish()` to stream the output, otherwise the program exits on the console while continuing to run in the background, causing the assertions in the test to fail due to empty input.

Reply:
This is a streaming job so the job will never end unless manually ended, so I think without `waitUntilFinish` is better here.

Reply:
are you able to see the printouts in console without `.waitUntilFinish()` during tests? i didn't seem able to but not sure why. adding it allows me to observe logs in cloud console. 

![Screenshot 2023-07-07 at 12 13 30 PM](https://github.com/GoogleCloudPlatform/java-docs-samples/assets/9518298/930a6558-a556-4110-b4d9-5c4eebc3974d)